pkg/registry/populator_test.go

```diff
@@ -1534,3 +1534,78 @@ func TestSemverPackageManifest(t *testing.T) {
 		})
 	}
 }
+
+func TestSubstitutesFor(t *testing.T) {
+	logrus.SetLevel(logrus.DebugLevel)
+	db, cleanup := CreateTestDb(t)
+	defer cleanup()
+
+	query, err := createAndPopulateDB(db)
+	require.NoError(t, err)
+
+	// add substitutesfor prometheus.0.15.0
+	substitutesfor := "prometheus.0.15.0.substitutesfor"
+	load, err := sqlite.NewSQLLiteLoader(db)
+	require.NoError(t, err)
+	graphLoader, err := sqlite.NewSQLGraphLoaderFromDB(db)
+	require.NoError(t, err)
+	p := registry.NewDirectoryPopulator(
+		load,
+		graphLoader,
+		query,
+		map[image.Reference]string{
+			image.SimpleReference("quay.io/test/" + substitutesfor): "../../bundles/" + substitutesfor},
+		make(map[string]map[image.Reference]string, 0), false)
+	require.NoError(t, p.Populate(registry.ReplacesMode))
+
+	// Check graph is unchanged but has new csv name + version
+	// Ensure bundlePaths in db match
+	bundles, err := query.ListBundles(context.Background())
+	expectedBundlePaths := []string{
+		"quay.io/test/prometheus.0.22.2",
+		"quay.io/test/prometheus.0.14.0",
+		"quay.io/test/prometheus.0.15.0.substitutesfor",
+		"quay.io/test/prometheus.0.15.0.substitutesfor",
+		"quay.io/test/prometheus.0.15.0",
+		"quay.io/test/prometheus.0.15.0",
+		"quay.io/test/etcd.0.9.0",
+		"quay.io/test/etcd.0.9.0",
+		"quay.io/test/etcd.0.9.0",
+		"quay.io/test/etcd.0.9.2",
+		"quay.io/test/etcd.0.9.2",
+		"quay.io/test/prometheus.0.14.0",
+	}
+	expectedChannels := []string{
+		"alpha",
+		"stable",
+		"stable",
+		"beta",
+		"stable",
+		"preview",
+		"preview",
+		"stable",
+		"stable",
+		"preview",
+		"preview",
+		"alpha",
+	}
+	require.NoError(t, err)
+	var bundlePaths []string
+	var channels []string
+	for _, bundle := range bundles {
+		bundlePaths = append(bundlePaths, bundle.BundlePath)
+		channels = append(channels, bundle.ChannelName)
+	}
+	require.ElementsMatch(t, expectedBundlePaths, bundlePaths)
+	require.ElementsMatch(t, expectedChannels, channels)
+
+	// Check channel head is unchanged for preview channel
+	head, err := query.GetCurrentCSVNameForChannel(context.Background(), "prometheus", "preview")
+	require.NoError(t, err)
+	require.Equal(t, "prometheusoperator.0.22.2", head)
+
+	// Check channel head is changed for stable channel
+	head, err = query.GetCurrentCSVNameForChannel(context.Background(), "prometheus", "stable")
+	require.NoError(t, err)
+	require.Equal(t, "prometheusoperator.0.15.0+1-freshmaker-rebuild", head)
+}
```

Comment:
Note: still need more tests for more complex cases and more thorough graph checking